### PR TITLE
MessageBuilder update, comments welcome

### PR DESCRIPTION
**Pull request by Almighty-Alpaca**
:eyes:

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 1]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                switch (mention)
+                {
+                    case EVERYONE:
+                        replaceAll("@everyone", "@\u200Beveryone");
+                        break;
+                    case HERE:
+                        replaceAll("@here", "@\u200Bhere");
+                        break;
+                    case CHANNEL:
+                        {
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Brackets should be in-line with the case statement.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 2]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                            {
+                                TextChannel channel = jda.getTextChannelById(matcher.group(1));
+                                if (channel != null)
+                                {
+                                    
+                                    replaceAll(matcher.group(), "#" + channel.getName());
+                                }
+                            }
+                        }
+                        break;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: put the break inside of the case brace scope for clarity

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 3]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                                    Role role = g.getRoleById(matcher.group(1));
+                                    if (role != null)
+                                    {
+                                        replaceAll(matcher.group(), "@"+role.getName());
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        break;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Same issues here with brace and break.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 4]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                                    if (guild != null && (member = guild.getMember(user)) != null)
+                                        replacement = member.getEffectiveName();
+                                    else
+                                        replacement = user.getName();
+
+                                    replaceAll(matcher.group(), "@" + replacement);
+                                }
+                            }
+                        }
+                        break;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: and here.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 5]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        
+        return this;
+    }
+
+    /**
+     * Returns the underlying {@link StringBuilder}.
+     * 
+     * @return the {@link StringBuilder} used by this {@link MessageBuilder}
+     */
+    public StringBuilder getStringBuilder() {
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: newline the bracket.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 6]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        if (targetCount == 0)
+        {
+            return fromIndex;
+        }
+
+        char strFirstChar = target.charAt(0);
+
+        int max = endIndex + targetCount - 1;
+
+        lastCharSearch: for (int i = fromIndex; i <= max; i++)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: put the label above the forloop

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 7]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        }
+
+        int targetCount = target.length();
+        if (targetCount == 0)
+        {
+            return fromIndex;
+        }
+
+        char strFirstChar = target.charAt(0);
+
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: unnecessary whitespace between the local variables. Remove this line.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 8]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        {
+            fromIndex = rightIndex;
+        }
+
+        int strLastIndex = targetCount - 1;
+        char strLastChar = target.charAt(strLastIndex);
+
+        int min = fromIndex + targetCount - 1;
+
+        lastCharSearch: for (int i = endIndex; i >= min; i--)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Put label above forloop

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 9]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+     * The method will try the policies in the order they are passed to it.
+     * <p>
+     * This is not MarkDown safe. An easy workaround is to include Zero Witdh Spaces as predetermined breaking points to the message and only split on them.
+     * 
+     * @return the created {@link net.dv8tion.jda.core.entities.Message Messages}
+     */
+    public Queue<Message> buildAll(SplitPolicy... policy)
+    {
+        if (builder.length() == 0)
+            throw new UnsupportedOperationException("Cannot build a Message with no content. (You never added any content to the message)");
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: The possibility of an embed only message is never considered. Consider `this.isEmpty()`

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 10]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        } 
+
+        if (policy.length == 0)
+        {
+            policy = new SplitPolicy[]{ SplitPolicy.ANYWHERE };
+        }
+        
+        int currentBeginIndex = 0;
+
+        messageLoop: while (currentBeginIndex < builder.length() - 2001)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Place label above loop.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 11]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+            for (int i = 0; i < policy.length; i++)
+            {
+                try
+                {
+                    int currentEndIndex = policy[i].nextMessage(currentBeginIndex, this);
+                    messages.add(build(currentBeginIndex, currentEndIndex));
+                    currentBeginIndex = currentEndIndex;
+                    continue messageLoop;
+                }
+                catch (Exception e) {}
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: `e` -> `ignored`
That being said, what possible exceptions are you guarding against here? @Almighty-Alpaca

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 12]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                {
+                    throw new IllegalArgumentException("could not split the message");
+                }
+                else
+                {
+                    currentEndIndex += this.chars.length();
+                    return currentEndIndex;
+                }
+            }
+
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: unneeded newlines

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 13]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        /**
+         * Splits on newline chars <code>\n</code>.
+         */
+        public static final SplitPolicy NEWLINE = new CharSequenceSplitPolicy("\n", true);
+
+        /**
+         * Splits on space chars.
+         */
+        public static final SplitPolicy SPACE = new CharSequenceSplitPolicy(" ", true);
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Place constants at the top of the Interface.

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 14]
```diff
@@ -398,6 +950,6 @@ private String getTag()
         {
             return tag;
         }
-
     }
+
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: Unneeded newline

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 1 above)
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: why even use the brackets here?

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 15]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+                {
+                    currentEndIndex += this.chars.length();
+                    return currentEndIndex;
+                }
+            }
+
+        }
+
+        /**
+         * Splits on newline chars <code>\n</code>.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: `{@code \n}` > `<code>\n</code>`

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 16]
```diff
@@ -366,18 +431,505 @@ public int getLength()
     public Message build()
     {
         String message = builder.toString();
-        if (message.isEmpty() && embed == null)
+        if (message.isEmpty())
```
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 17]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+
+    /**
+     * Removes all mentions and replaces them with the closest looking textual representation.<p>
+     * <p>
+     * Use this over {@link #stripMentions(Guild)} if {@link MentionType.USER User} mentions should be replaced with their user names
+     *
+     * @param jda the JDA instance
+     * @return this instance
+     */
+    public MessageBuilder stripMentions(JDA jda) {
```
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 18]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+
+    /**
+     * Removes all mentions and replaces them with the closest looking textual representation.<p>
+     * <p>
+     * Use this over {@link #stripMentions(JDA)} if {@link MentionType.USER User} mentions should be replaced with their nicknames in a specific guild
+     *
+     * @param guild the guild for {@link MentionType.USER User} mentions 
+     * @return this instance
+     */
+    public MessageBuilder stripMentions(Guild guild) {
```
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 19]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+     * @param types the {@link MentionType MentionTypes} that should be stripped
+     * @return this instance
+     */
+    public MessageBuilder stripMentions(JDA jda, MentionType... types) {
+        return this.stripMentions(jda, (Guild) null, types);
+    }
+
+    private MessageBuilder stripMentions(JDA jda, Guild guild, MentionType... types) {
+        String string = null;
+        if (types != null)
```
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: if yo just call `#stripMentions(jda, guild, null)` the `types` argument will be null, not an empty array

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java [hunk 20]
```diff
@@ -366,18 +431,505 @@ public int getLength()
...
+        {
+            for (MentionType mention : types)
```
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 6 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 8 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 9 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: forgot to change to `this.isEmpty()` because I made that method after the build() methods

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 12 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 11 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: `SplitPolicy#nextMessage()` is supposed to throw an exception if finding the nest message's end fails (for example if there's no newline char within a range of 2k chars). looking at it know I think it would be better if it just returns -1 in such a case @DV8FromTheWorld

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 2 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: solved with removing the braces :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 3 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: same as above :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 4 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: same as above :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 5 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 7 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 10 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 15 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:

File: src/main/java/net/dv8tion/jda/core/MessageBuilder.java (hunk 13 above)
Comment (Almighty-Alpaca) [inline src/main/java/net/dv8tion/jda/core/MessageBuilder.java]: :+1:
